src/main/java/com/android/volley/toolbox/AsyncHttpStack.java

```diff
@@ -0,0 +1,113 @@
+     *
+     * @param request to perform
+     * @param additionalHeaders to be sent together with {@link Request#getHeaders()}
+     * @param callback to be called after retrieving the {@link HttpResponse} or throwing an error.
+     */
+    public abstract void executeRequest(
+            Request<?> request, Map<String, String> additionalHeaders, OnRequestComplete callback);
+
+    /**
+     * Performs an HTTP request with the given parameters.
+     *
+     * @param request the request to perform
+     * @param additionalHeaders additional headers to be sent together with {@link
+     *     Request#getHeaders()}
+     * @return the {@link HttpResponse}, or null if an InterruptedException occurs.
+     * @throws IOException if an I/O error occurs during the request
+     * @throws AuthFailureError if an authentication failure occurs during the request
+     */
+    @Nullable
+    @Override
+    public final HttpResponse executeRequest(
+            Request<?> request, Map<String, String> additionalHeaders)
+            throws IOException, AuthFailureError {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicReference<Response> entry = new AtomicReference<>();
+        executeRequest(
+                request,
+                additionalHeaders,
+                new OnRequestComplete() {
+                    @Override
+                    public void onSuccess(HttpResponse httpResponse) {
+                        Response response = new Response(httpResponse, null, null, 1);
+                        entry.set(response);
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public void onAuthError(AuthFailureError authFailureError) {
+                        Response response = new Response(null, null, authFailureError, 3);
+                        entry.set(response);
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public void onError(IOException ioException) {
+                        Response response = new Response(null, ioException, null, 2);
+                        entry.set(response);
+                        latch.countDown();
+                    }
+                });
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            VolleyLog.e(e, "while waiting for CountDownLatch");
+            Thread.currentThread().interrupt();
+            return null;
+        }
+        int type = entry.get().type;
+        if (type == 1) {
+            return entry.get().httpResponse;
+        } else if (type == 2) {
+            throw entry.get().ioException;
+        } else {
+            throw entry.get().authFailureError;
+        }
+    }
+
+    private static class Response {
+        HttpResponse httpResponse;
+        IOException ioException;
+        AuthFailureError authFailureError;
+        int type;
```

Comment:
You could use an @IntDef for this to get better type safety without the cost of an enum (which would normally be preferred), but then I'm not sure why you need this field at all. Can't you just do:

```java
Response response = entry.get();
if (response.httpResponse != null) {
  return response.httpResponse;
} else if (response.ioException != null) {
  ...
}
...
```

Reply:
Yeah, that works. I thought having a type parameter would make things simpler, but I think it obscures it.

Reply:
we do something similar in RequestFuture